File: test-automation/python/pages/login_page.py
"""
Page Object Model for SecureDocs Login Page
Covers test cases: AUTH_001 to AUTH_010
"""

from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import Select
from selenium.common.exceptions import TimeoutException, NoSuchElementException
from .base_page import BasePage


class LoginPage(BasePage):
    """Login page object model class"""
    
    # Page URL
    LOGIN_URL = "/login"
    
    # Locators
    EMAIL_FIELD = (By.ID, "email")
    PASSWORD_FIELD = (By.ID, "password")
    LOGIN_BUTTON = (By.CSS_SELECTOR, "button[type='submit']")
    REGISTER_LINK = (By.LINK_TEXT, "Register")
    FORGOT_PASSWORD_LINK = (By.LINK_TEXT, "Forgot Your Password?")
    ERROR_MESSAGE = (By.CSS_SELECTOR, ".alert-danger")
    SUCCESS_MESSAGE = (By.CSS_SELECTOR, ".alert-success")
    REMEMBER_ME_CHECKBOX = (By.ID, "remember")
    LANGUAGE_SELECTOR = (By.CSS_SELECTOR, "select[name='language']")
    WEBAUTHN_LOGIN_BTN = (By.ID, "webauthn-login-btn")
    VERIFICATION_NOTICE = (By.CSS_SELECTOR, ".verification-notice")
    
    def __init__(self, driver):
        super().__init__(driver)
        
    def navigate_to_login(self):
        """Navigate to login page"""
        self.driver.get(self.base_url + self.LOGIN_URL)
        self.wait_for_page_load()
        
    def wait_for_page_load(self):
        """Wait for login page to load"""
        self.wait_for_element_visible(self.EMAIL_FIELD)
        
    def login(self, email, password):
        """
        Perform login with credentials
        Test Cases: AUTH_001, AUTH_002, AUTH_003, AUTH_004
        """
        self.clear_and_type(self.EMAIL_FIELD, email)
        self.clear_and_type(self.PASSWORD_FIELD, password)
        self.click_element(self.LOGIN_BUTTON)
        
    def login_with_remember_me(self, email, password):
        """
        Perform login with remember me option
        Test Case: AUTH_001 (variant)
        """
        self.clear_and_type(self.EMAIL_FIELD, email)
        self.clear_and_type(self.PASSWORD_FIELD, password)
        
        remember_checkbox = self.find_element(self.REMEMBER_ME_CHECKBOX)
        if not remember_checkbox.is_selected():
            self.click_element(self.REMEMBER_ME_CHECKBOX)
            
        self.click_element(self.LOGIN_BUTTON)
        
    def attempt_login_with_empty_fields(self):
        """
        Attempt login with empty fields
        Test Case: AUTH_004
        """
        self.click_element(self.LOGIN_BUTTON)
        
    def click_register_link(self):
        """
        Navigate to registration page
        Test Case: AUTH_005
        """
        self.click_element(self.REGISTER_LINK)
        
    def click_forgot_password_link(self):
        """
        Navigate to password reset page
        Test Case: AUTH_009
        """
        self.click_element(self.FORGOT_PASSWORD_LINK)
        
    def switch_language(self, language):
        """
        Switch language
        Related to: DASH_004
        """
        language_select = Select(self.find_element(self.LANGUAGE_SELECTOR))
        language_select.select_by_visible_text(language)
        
    def click_webauthn_login(self):
        """
        Initiate WebAuthn login
        Test Cases: WEBAUTH_002
        """
        self.click_element(self.WEBAUTHN_LOGIN_BTN)
        
    # Validation Methods
    
    def is_login_successful(self, timeout=10):
        """
        Check if login was successful by verifying redirect
        Test Case: AUTH_001
        """
        try:
            WebDriverWait(self.driver, timeout).until(
                EC.url_contains("/dashboard")
            )
            return True
        except TimeoutException:
            return False
            
    def get_error_message(self, timeout=5):
        """
        Get error message text
        Test Cases: AUTH_002, AUTH_003, AUTH_004
        """
        try:
            error_element = WebDriverWait(self.driver, timeout).until(
                EC.visibility_of_element_located(self.ERROR_MESSAGE)
            )
            return error_element.text
        except TimeoutException:
            return ""
            
    def get_success_message(self, timeout=5):
        """
        Get success message text
        Test Case: AUTH_005
        """
        try:
            success_element = WebDriverWait(self.driver, timeout).until(
                EC.visibility_of_element_located(self.SUCCESS_MESSAGE)
            )
            return success_element.text
        except TimeoutException:
            return ""
            
    def is_email_verification_notice_visible(self):
        """
        Check if email verification notice is displayed
        Test Case: OTP_006
        """
        try:
            return self.find_element(self.VERIFICATION_NOTICE).is_displayed()
        except NoSuchElementException:
            return False
            
    def has_validation_errors(self):
        """
        Check if login form has validation errors
        Test Case: AUTH_004
        """
        try:
            email_error = self.driver.find_element(By.CSS_SELECTOR, "#email + .error-message")
            password_error = self.driver.find_element(By.CSS_SELECTOR, "#password + .error-message")
            return email_error.is_displayed() or password_error.is_displayed()
        except NoSuchElementException:
            return False
            
    def get_field_validation_error(self, field_id):
        """
        Verify specific field validation error
        Test Case: AUTH_004
        """
        try:
            error_element = self.driver.find_element(
                By.CSS_SELECTOR, f"#{field_id} + .error-message"
            )
            return error_element.text
        except NoSuchElementException:
            return ""
            
    def is_webauthn_available(self):
        """
        Check if WebAuthn is supported/available
        Test Case: WEBAUTH_004
        """
        try:
            return self.find_element(self.WEBAUTHN_LOGIN_BTN).is_displayed()
        except NoSuchElementException:
            return False

File: test-automation/sample-pom/LoginPage.java
package com.securedocs.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.support.ui.ExpectedConditions;
import java.time.Duration;

/**
 * Page Object Model for SecureDocs Login Page
 * Covers test cases: AUTH_001 to AUTH_010
 */
public class LoginPage {
    
    private WebDriver driver;
    private WebDriverWait wait;
    
    // Page URL
    private static final String LOGIN_URL = "/login";
    
    // Page Elements using @FindBy annotations
    @FindBy(id = "email")
    private WebElement emailField;
    
    @FindBy(id = "password") 
    private WebElement passwordField;
    
    @FindBy(css = "button[type='submit']")
    private WebElement loginButton;
    
    @FindBy(linkText = "Register")
    private WebElement registerLink;
    
    @FindBy(linkText = "Forgot Your Password?")
    private WebElement forgotPasswordLink;
    
    @FindBy(css = ".alert-danger")
    private WebElement errorMessage;
    
    @FindBy(css = ".alert-success")
    private WebElement successMessage;
    
    @FindBy(id = "remember")
    private WebElement rememberMeCheckbox;
    
    // Language switching elements
    @FindBy(css = "select[name='language']")
    private WebElement languageSelector;
    
    // WebAuthn login button
    @FindBy(id = "webauthn-login-btn")
    private WebElement webAuthnLoginButton;
    
    // Email verification elements
    @FindBy(css = ".verification-notice")
    private WebElement verificationNotice;
    
    // Constructor
    public LoginPage(WebDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        PageFactory.initElements(driver, this);
    }
    
    // Navigation Methods
    public void navigateToLogin() {
        driver.get(System.getProperty("base.url", "http://localhost:8000") + LOGIN_URL);
        waitForPageLoad();
    }
    
    public void waitForPageLoad() {
        wait.until(ExpectedConditions.presenceOfElementLocated(By.id("email")));
    }
    
    // Authentication Actions
    
    /**
     * Perform login with credentials
     * Test Cases: AUTH_001, AUTH_002, AUTH_003, AUTH_004
     */
    public void login(String email, String password) {
        clearAndType(emailField, email);
        clearAndType(passwordField, password);
        loginButton.click();
    }
    
    /**
     * Perform login with remember me option
     * Test Case: AUTH_001 (variant)
     */
    public void loginWithRememberMe(String email, String password) {
        clearAndType(emailField, email);
        clearAndType(passwordField, password);
        if (!rememberMeCheckbox.isSelected()) {
            rememberMeCheckbox.click();
        }
        loginButton.click();
    }
    
    /**
     * Attempt login with empty fields
     * Test Case: AUTH_004
     */
    public void attemptLoginWithEmptyFields() {
        loginButton.click();
    }
    
    /**
     * Navigate to registration page
     * Test Case: AUTH_005
     */
    public void clickRegisterLink() {
        registerLink.click();
    }
    
    /**
     * Navigate to password reset page  
     * Test Case: AUTH_009
     */
    public void clickForgotPasswordLink() {
        forgotPasswordLink.click();
    }
    
    /**
     * Switch language
     * Related to: DASH_004
     */
    public void switchLanguage(String language) {
        languageSelector.sendKeys(language);
    }
    
    /**
     * Initiate WebAuthn login
     * Test Cases: WEBAUTH_002
     */
    public void clickWebAuthnLogin() {
        webAuthnLoginButton.click();
    }
    
    // Validation Methods
    
    /**
     * Check if login was successful by verifying redirect
     * Test Case: AUTH_001
     */
    public boolean isLoginSuccessful() {
        try {
            wait.until(ExpectedConditions.urlContains("/dashboard"));
            return true;
        } catch (Exception e) {
            return false;
        }
    }
    
    /**
     * Get error message text
     * Test Cases: AUTH_002, AUTH_003, AUTH_004
     */
    public String getErrorMessage() {
        try {
            wait.until(ExpectedConditions.visibilityOf(errorMessage));
            return errorMessage.getText();
        } catch (Exception e) {
            return "";
        }
    }
    
    /**
     * Get success message text
     * Test Case: AUTH_005
     */
    public String getSuccessMessage() {
        try {
            wait.until(ExpectedConditions.visibilityOf(successMessage));
            return successMessage.getText();
        } catch (Exception e) {
            return "";
        }
    }
    
    /**
     * Check if email verification notice is displayed
     * Test Case: OTP_006
     */
    public boolean isEmailVerificationNoticeVisible() {
        try {
            return verificationNotice.isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }
    
    /**
     * Check if login form has validation errors
     * Test Case: AUTH_004
     */
    public boolean hasValidationErrors() {
        try {
            WebElement emailError = driver.findElement(By.cssSelector("#email + .error-message"));
            WebElement passwordError = driver.findElement(By.cssSelector("#password + .error-message"));
            return emailError.isDisplayed() || passwordError.isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }
    
    /**
     * Verify specific field validation error
     * Test Case: AUTH_004
     */
    public String getFieldValidationError(String fieldId) {
        try {
            WebElement errorElement = driver.findElement(By.cssSelector("#" + fieldId + " + .error-message"));
            return errorElement.getText();
        } catch (Exception e) {
            return "";
        }
    }
    
    /**
     * Check if WebAuthn is supported/available
     * Test Case: WEBAUTH_004
     */
    public boolean isWebAuthnAvailable() {
        try {
            return webAuthnLoginButton.isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }
    
    // Utility Methods
    
    /**
     * Clear field and type new value
     */
    private void clearAndType(WebElement element, String text) {
        element.clear();
        element.sendKeys(text);
    }
    
    /**
     * Wait for element to be clickable
     */
    private void waitForClickable(WebElement element) {
        wait.until(ExpectedConditions.elementToBeClickable(element));
    }
    
    /**
     * Get current page title
     */
    public String getPageTitle() {
        return driver.getTitle();
    }
    
    /**
     * Get current URL
     */
    public String getCurrentUrl() {
        return driver.getCurrentUrl();
    }
    
    /**
     * Check if element is visible
     */
    public boolean isElementVisible(WebElement element) {
        try {
            return element.isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }
}
